lib/features/docker_save.js

```diff
@@ -9,41 +11,100 @@ import zlib from 'zlib';
 
 let debug = Debug('docker-worker:features:docker-save');
 
-const ARTIFACT_NAME = 'public/dockerImage.tar'
+//must be changed if the delimiter in lib/volume_cache.js is changed
+let KEY_DELIMITER = '::';
 
 export default class DockerSave {
   async killed (task) {
-    //temporary path for saved file
-    let pathname = path.join('/tmp', slugid.v4() + '.tar.gz');
-
-    let {Id: imageId} = await task.dockerProcess.container.commit({
-      repo: 'task/' + task.status.taskId + '/' + task.runId
-    });
-    let image = task.runtime.docker.getImage('task/' + task.status.taskId + '/' + task.runId + ':latest');
-    let imgStream = await image.get();
-    let zipStream = zlib.createGzip();
-    imgStream.pipe(zipStream).pipe(fs.createWriteStream(pathname));
-    await waitForEvent(zipStream, 'end');
-    debug('tar written');
-
-    let stat = await fs.stat(pathname);
-    debug(stat.size);
-    let uploadStream = fs.createReadStream(pathname);
-    let expiration = new Date(Date.now() + task.runtime.dockerSave.expiration);
-
-    await uploadToS3(task, uploadStream, ARTIFACT_NAME, expiration, {
-      'content-type': 'application/x-tar',
-      'content-length': stat.size,
-      'content-encoding': 'gzip'
-    });
-
-    debug('artifact uploaded');
-
-    //cleanup
-    fs.unlink(pathname).catch(() => {
-      task.runtime.log('[alert-operator] could not delete docker save tarball, worker may run out of hdd space');
-    });
-    await image.remove();
+    let errors = [];
+    let volumeCaches = task.volumeCaches || [];
+    await Promise.all(volumeCaches.map(async (cacheStr) => {
```

Comment:
Small thing... but if you have two min.. maybe refactor into two methods:
```js
 await Promise.all(
  volumeCaches
    .map(cacheStr => this.uploadCache(task, cacheStr).catch(err => error.push(err)))
    .concat(this.uploadContainer(task).catch(err => error.push(err)))
)
```
Then you won't need the `try/catch` inside either `uploadCache` or `uploadContainer` and the code will be a little easier to read.. Especially, if you write a comment documenting the two methods like:
```js
  /** Upload a cache folder as an artifact */
  async uploadCache(task, cacheStr) {
    ....
  },
  /** Commit and upload task container as an artifact */
  async uploadContainer(task) {
    ....
  }
```